src/main/java/org/jenkinsci/plugins/awsdevicefarm/AWSDeviceFarmRecorder.java

```diff
@@ -557,6 +579,46 @@ public void perform(@Nonnull hudson.model.Run<?, ?> build, @Nonnull FilePath wor
             writeToLog(log, String.format("Using Project '%s'", projectName));
             Project project = adf.getProject(projectName);
 
+            // Device Selection logic
+            DeviceSelectionConfiguration deviceSelectionConfig = null;
+            ArrayList<DeviceFilter> deviceFilterList = null;
+            DeviceFilter deviceFilter = null;
+
+            if (deviceSelection != null && !deviceSelection.isEmpty()) {
+                deviceSelectionConfig = new DeviceSelectionConfiguration();
+                deviceFilter = new DeviceFilter();
+                deviceFilterList = new ArrayList<DeviceFilter>();
+
+                String deviceSelectionStr = null;
+                deviceSelectionStr = String.format("%s", env.expand(deviceSelection));
+                String[] deviceSelectionList = deviceSelectionStr.split(",");
+                writeToLog(log, String.format("Using Device Selection Filters '%s'", Arrays.toString(deviceSelectionList)));
+
+                for (String deviceFilterString : deviceSelectionList) {
+                    String[] filterElements = deviceFilterString.split(" ", 3);
+                    if (filterElements.length != 3) {
+                        writeToLog(log, String.format("Invalid Device Selection Filter '%s'", deviceFilterString));
+                        build.setResult(Result.FAILURE);
+                        return;
+                    }
+                    deviceFilter  = new DeviceFilter();
+                    deviceFilter.setAttribute(filterElements[0]);
+                    deviceFilter.setOperator(filterElements[1]);
+                    ArrayList<String> testValues = new ArrayList<String>();
+                    testValues.add(filterElements[2]);
+                    deviceFilter.setValues(testValues);
+                    deviceFilterList.add(deviceFilter);
+                    writeToLog(log, String.format("Device Selection Filter List '%s'", deviceFilterList.toString()));
+                }
+
+
+                writeToLog(log, String.format("Updated Device Filters '%s'", deviceSelectionConfig.toString()));
+                deviceSelectionConfig.setFilters(deviceFilterList);
+                if(maxDevicesForDeviceSelection != null && maxDevicesForDeviceSelection > 0) {
```

Comment:
Is maxDevicesForDeviceSelection mandatory argument? If so you will need to consider adding validation in the beginning

Reply:
You are right. Seems like it is when you use deviceSelectionConfig but not when you use devicePool
https://docs.aws.amazon.com/devicefarm/latest/APIReference/API_DevicePool.html

I went with the thinking they both are the same. Will change this.

Reply:
I added a check at that start with the deviceSelectionConfig and one before scheduleRun to throw an IllegalArgument exception.